app/src/main/java/com/example/countryinfo/CountryDetailsAdapter.kt

```diff
@@ -0,0 +1,162 @@
+                    bindMultiple(DetailsStates.LANGUAGES, countryDetails)
+                }
+                6 -> {
+                    itemView.labelText.text = DetailsStates.TIMEZONES.title
+                    bindMultiple(DetailsStates.TIMEZONES, countryDetails)
+                }
+                7 -> {
+                    itemView.labelText.text = DetailsStates.CALLING_CODES.title
+                    bindMultiple(DetailsStates.CALLING_CODES, countryDetails)
+                }
+            }
+        }
+
+        private fun bindMultiple(state: DetailsStates, countryDetails: CountryDetails) {
+            if (state == DetailsStates.values().last()) {
+                itemView.lineImage.visibility = View.GONE
+            }
+            when (state) {
+                DetailsStates.CURRENCIES -> {
+                    setTextViews(countryDetails.currencyNames, state)
+                    dialogMessage = countryDetails.currencyNames
+                }
+                DetailsStates.LANGUAGES -> {
+                    setTextViews(countryDetails.languages, state)
+                    dialogMessage = countryDetails.languages
+                }
+                DetailsStates.TIMEZONES -> {
+                    setTextViews(countryDetails.timezones, state)
+                    dialogMessage = countryDetails.timezones
+                }
+                DetailsStates.CALLING_CODES -> {
+                    setTextViews(countryDetails.callingCodes, state)
+                    dialogMessage = countryDetails.callingCodes
+                }
+            }
+            viewState = state
+        }
+
+        private fun setTextViews(values: List<String>, state: DetailsStates) {
+            for (i in values.indices) {
+                if (i > 2)
+                    break
+                when (i) {
+                    0 -> {
+                        itemView.titleText.text = values[i]
+                        itemView.titleText.setBackgroundResource(state.bgShapeId)
+                    }
+                    1 -> {
+                        itemView.titleText2.text = values[i]
+                        itemView.titleText2.setBackgroundResource(state.bgShapeId)
+                        itemView.titleText2.visibility = View.VISIBLE
+                    }
+                    2 -> {
+                        itemView.titleText3.text = values[i]
+                        itemView.titleText3.setBackgroundResource(state.bgShapeId)
+                        itemView.titleText3.visibility = View.VISIBLE
+                    }
+                }
+            }
+        }
+
+        private fun formatPopulation(value: Double): String {
+            val format = DecimalFormat("0.##")
+            val res = if (value >= 1_000_000) {
+                format.format(value / 1_000_000) + " m"
+            } else {
+                format.format(value)
+            }
+            return res.replace(".", ",")
+        }
```

Comment:
[nit] This should be done in the ViewModel layer.

Reply:
Changed.